rust/src/server.rs

```diff
@@ -0,0 +1,736 @@
+use std::collections::BTreeMap;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use futures::future::{select, Either};
+use futures::StreamExt;
+use tokio::sync::{broadcast, mpsc, oneshot, Mutex, RwLock};
+use tonic::{Request, Response};
+
+use crate::prelude::*;
+
+use proto::seabird::seabird_server::{Seabird, SeabirdServer};
+use proto::{
+    seabird::chat_ingest_server::{ChatIngest, ChatIngestServer},
+    ChatEventInner, EventInner,
+};
+
+const CHAT_INGEST_SEND_BUFFER: usize = 10;
+const CHAT_INGEST_RECEIVE_BUFFER: usize = 10;
+const BROADCAST_BUFFER: usize = 32;
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(1);
+
+#[derive(Debug)]
+struct ChatBackend {
+    id: BackendId,
+    receiver: mpsc::Receiver<proto::ChatRequest>,
+    sender: broadcast::Sender<proto::Event>,
+    cleanup: mpsc::UnboundedSender<BackendId>,
+}
+
+impl Drop for ChatBackend {
+    fn drop(&mut self) {
```

Comment:
N00b question: if you `impl Drop` here, will all of the objects you've defined still get `Drop` called for them or do you need to clean them up here?

Reply:
Yeah, all the objects will still get Drop called.